### PR TITLE
Rebase chitoku/jp4.6_tensorrt8 onto latest master. Make it backwards compatible for TRT v6.x-8.x.

### PR DESCRIPTION
**Pull request by chaoz-dev**
- Update plugin classes to match the base IPluginV2 class in TensorRT8
- Cherry-pick from gcunhase/torch2trt
- Cherry-pick from gcunhase/torch2trt
- Cherry-pick from gcunhase/torch2trt
- Fix merge errors

### CODE REVIEW

**Comment by chaoz-dev** [comment]
@chitoku I can try merging this onto your branch, but I can't seem to push to your fork?

**Comment by chaoz-dev** [comment]
Note: Running `sudo python3 setup.py install --plugins`, these changes will fail for TRT 6.0.1 and 7.2.3.4 since `noexcept` has not yet been added for several APIs yet until TRT 8.

File: setup.py [hunk 1]
```diff
@@ -12,8 +13,14 @@ def trt_lib_dir():
 ext_modules = []
 exclude_dir = ["torch2trt/contrib","torch2trt/contrib.*"]
 
+compile_args_cxx = []
+if torch.__version__ < '1.5':
+    compile_args_cxx.append('-DUSE_DEPRECATED_INTLIST')
+if tensorrt.__version__ < '8':
+    compile_args_cxx.append('-DPRE_TRT8')
```
Comment (chaoz-dev) [inline setup.py:20]: Alternate flag names welcomed here.

**Comment by chaoz-dev** [comment]
Updated the code to be backwards compatible; now it should work for TRT versions 6-8.

**Comment by jaybdub** [comment]
Hi Chaoz-dev.

Thanks for looking into the backwards compatability.

I'm running test cases against various environments now.


The NGC 21-06 container worked well, but for the NGC 21-09 PyTorch container, I'm running into build errors, seemingly related to the PyTorch version (1.10.0a0+3fd9dcf).  Likely this issue already exists today anyways, so perhaps not a blocker for merging.

Do you mind sharing which version of PyTorch you tested against while testing support for TensorRT 8?

Best,
John

**Comment by jaybdub** [comment]
Resolved this issue, the issue was with semantic version checking in setup.py.  '1.10' was falsely registered as < '1.5'.

Using 

```
from packaging import version

version('1.10') < version('1.5')
```

Orders the versions correctly, and it builds for later versions of PyTorch.  

Now running into following issue while testing group norm plugin with TensorRT 8.0

```bash
TypeError: add_plugin_v2(): incompatible function arguments. The following argument types are supported:
    1. (self: tensorrt.tensorrt.INetworkDefinition, inputs: List[tensorrt.tensorrt.ITensor], plugin: tensorrt.tensorrt.IPluginV2) -> tensorrt.tensorrt.IPluginV2Layer

Invoked with: <tensorrt.tensorrt.INetworkDefinition object at 0x7f5bb4f402b0>, [<tensorrt.tensorrt.ITensor object at 0x7f5bb57a10f0>], <torch2trt.plugins.GroupNormPlugin object at 0x7f5bb6428cf0>
```

Seems like the TensorRT python API isn't happy with the arguments provided to add_plugin_v2. Will look into this, but curious if you experience this with your setup?

Best,
John

**Comment by chaoz-dev** [comment]
Hey John,

I've tested this using the NGC 21-10 container, which is running TRT 8.0.3.4 and PyTorch 1.10.0a0+0aef44c. 

I've been able to build using this container without seeing the issues that you've mentioned... can I ask what commands you're running? I may have missed something here; I can try on my side to see if I can replicate your errors. 

Thanks!

Chao

**Comment by chaoz-dev** [comment]
Just gave this a try with the 21-09 NGC container, I'm getting some errors here while building. Let me take a look, and also see if this is present for the other recent containers.

**Comment by jaybdub** [comment]
Hi @chaoz-dev ,

Thanks for checking into this.

With the version checking fix in setup.py I mentioned above, I'm able to build using 21-10 container.

The error I've been unable to get past occurs at runtime when running cross validation tests that use plugins.  For example,

```bash
python3 -m torch2trt.test --name=group
```

Will run the group norm tests.

The issue seems to be that for some reason with the 21-10 container, the class returned by the following line 

https://github.com/NVIDIA-AI-IOT/torch2trt/blob/d68712af6dbcafb436efea91e49fd419ce561c90/torch2trt/converters/group_norm.py#L19

is a ``GroupNormPlugin`` class.  The ``add_plugin_v2`` method then rejects this data type.

With older container versions it returns the abstract ``IPluginV2`` type, which ``add_plugin_v2`` is happy with. 

I've been unable to figure this out yet, perhaps it has to do with pybind11 version as to why it's returning ``IPluginV2``.

As of now, it seems this will prevent us from adding any of the defined plugins to the network, which is a relatively large issue.

Look forward to hearing your thoughts or if your experience varies from what I saw.

Best,
John

**Comment by jaybdub** [comment]
This is merged now in master, with a small modification to handle version checking in setup.py.  The TRT Python API issue I described above likely still persists, so the plugins likely can't be used at runtime, but the setup.py build succeeds.  Thanks @chitoku and @chaoz-dev !

**Comment by kuu22** [comment]
Has there been any progress on this? Right now I can't use the GroupNorm plugin in TRT8. @jaybdub

**Pull request merged**